### PR TITLE
o/ifacestate: warn if the snapd.apparmor service is disabled

### PR DESCRIPTION
**Pull request by mardy**
If the snapd.apparmor service is not started, the AppArmor profiles of
snap-confine and snap-update-ns might not be loaded. If that happens, no
snap applications can start, as the calls to aa_change_onexec() will
fail.

We might want to take more drastic measures when this happens (like,
re-enabling the service ourselves?), but while we ponder the decision
emitting a warning seems like a harmless yet useful behavioural change.

See also: https://bugs.launchpad.net/snapd/+bug/1806135

### CODE REVIEW

File: overlord/ifacestate/ifacemgr.go [hunk 1]
```diff
@@ -174,6 +174,10 @@ func (m *InterfaceManager) StartUp() error {
 			return err
 		}
 	}
+	if snapdAppArmorServiceIsDisabled() {
```
Comment (bboozzoo) [inline overlord/ifacestate/ifacemgr.go:177]: I think this should be part of apparmor backend Initialize() which called only when apparmor support has been found in the kernel.

File: overlord/ifacestate/ifacemgr.go [hunk 2]
```diff
@@ -174,6 +174,10 @@ func (m *InterfaceManager) StartUp() error {
...
+		s.Warnf(`the snapd.apparmor service is disabled; snap applications will likely not start.
```
Comment (bboozzoo) [inline overlord/ifacestate/ifacemgr.go:178]: How about issue a similar warning from snap-confine? Looking at the manpage it should set errno to ENOENT when a profile does not exist

File: overlord/ifacestate/ifacemgr.go (hunk 1 above)
Comment (mardy) [reply overlord/ifacestate/ifacemgr.go:177]: That was my initial implementation, but from there we don't have access to the `state` and therefore cannot emit warnings. Samuele in the chat suggested to have the interface manager do this check.

File: overlord/ifacestate/ifacemgr.go (hunk 2 above)
Comment (mardy) [reply overlord/ifacestate/ifacemgr.go:178]: Good idea, I'll do some test.

Comment (mardy) [reply overlord/ifacestate/ifacemgr.go:178]: Yes, ENOENT is returned when the profiles are missing.

I've added a commit with your suggestion.

File: cmd/libsnap-confine-private/apparmor-support.c [hunk 3]
```diff
@@ -128,7 +128,14 @@ sc_maybe_aa_change_onexec(struct sc_apparmor *apparmor, const char *profile)
 		int aa_change_onexec_errno = errno;
 		if (secure_getenv("SNAPPY_LAUNCHER_INSIDE_TESTS") == NULL) {
 			errno = aa_change_onexec_errno;
-			die("cannot change profile for the next exec call");
+			if (errno == ENOENT) {
+				fprintf(stderr, "missing profile %s.\n"
+				        "Please make sure that the snapd.apparmor service is enabled\n",
```
Comment (bboozzoo) [inline cmd/libsnap-confine-private/apparmor-support.c:133]:
```suggestion
				        "Please make sure that the snapd.apparmor service is enabled and started\n",
```

File: cmd/snap-confine/snap-confine.c [hunk 4]
```diff
@@ -393,9 +393,11 @@ int main(int argc, char **argv)
 		// id is non-root.  This protects against, for example, unprivileged
 		// users trying to leverage the snap-confine in the core snap to
 		// escalate privileges.
+		errno = 0; // errno is insignificant here
 		die("snap-confine has elevated permissions and is not confined"
 		    " but should be. Refusing to continue to avoid"
-		    " permission escalation attacks");
+		    " permission escalation attacks\n"
+		    "Please make sure that the snapd.apparmor service is enabled.");
```
Comment (bboozzoo) [inline cmd/snap-confine/snap-confine.c:400]:
```suggestion
		    "Please make sure that the snapd.apparmor service is enabled and started.");
```

File: overlord/ifacestate/ifacemgr.go [hunk 5]
```diff
@@ -174,6 +174,10 @@ func (m *InterfaceManager) StartUp() error {
...
+Run "systemctl enable snapd.apparmor" to correct this.`)
```
Comment (bboozzoo) [inline overlord/ifacestate/ifacemgr.go:179]: The service needs to be started, so we should probably suggest to run this:
```suggestion
Run "systemctl enable --now snapd.apparmor" to correct this.`)
```

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**